src/html.rs

```diff
@@ -32,1559 +46,1730 @@ pub fn format_document_with_plugins<'a>(
+                    options,
+                    output,
+                    &$crate::Plugins::default(),
+                    Self::formatter,
+                )
+            }
 
-        let mut id = header.to_lowercase();
-        id = id
-            .chars()
-            .filter(is_permitted_char)
-            .map(|c| if c == ' ' { '-' } else { c })
-            .collect();
-
-        let mut uniq = 0;
-        id = loop {
-            let anchor = if uniq == 0 {
-                Cow::from(&id)
-            } else {
-                Cow::from(format!("{}-{}", id, uniq))
-            };
+            /// Formats an AST as HTML, modified by the given options. Accepts custom plugins.
+            #[inline]
+            pub fn format_document_with_plugins<'a, 'o, 'c: 'o>(
+                root: &'a $crate::nodes::AstNode<'a>,
+                options: &'o $crate::Options<'c>,
+                output: &'o mut dyn ::std::io::Write,
+                plugins: &'o $crate::Plugins<'o>,
+            ) -> ::std::io::Result<()> {
+                $crate::html::format_document_with_formatter(
+                    root,
+                    options,
+                    output,
+                    plugins,
+                    Self::formatter,
+                )
+            }
 
-            if !self.0.contains(&*anchor) {
-                break anchor.into_owned();
+            fn formatter<'a>(
+                context: &mut $crate::html::Context,
+                node: &'a $crate::nodes::AstNode<'a>,
+                entering: bool,
+            ) -> ::std::io::Result<::std::option::Option<$crate::html::RenderMode>> {
+                match node.data.borrow().value {
+                    $(
+                        $pat => {
+                            let mut suppress_children = false;
+                            $crate::formatter_captures!((context, node, entering, suppress_children), ($( $capture ),*));
+                            $case
+                            if suppress_children {
+                                Ok(::std::option::Option::None)
+                            } else {
+                                Ok(::std::option::Option::Some($crate::html::RenderMode::HTML))
+                            }
+                        }
+                    ),*
+                    _ => $crate::html::format_node_default(context, node, entering),
+                }
             }
+        }
+    };
+}
 
-            uniq += 1;
-        };
-        self.0.insert(id.clone());
-        id
-    }
+/// This must be exported so its uses in [`create_formatter!`] can be expanded,
+/// but it's not intended for direct use.
+#[doc(hidden)]
+#[macro_export]
+macro_rules! formatter_captures {
+    (($context:ident, $node:ident, $entering:ident, $suppress_children:ident), context, $bind:ident) => {
+        let $bind = $context;
+    };
+    (($context:ident, $node:ident, $entering:ident, $suppress_children:ident), output, $bind:ident) => {
```

Comment:
I wonder if we need both `output` and `context`. In order to call the `cr()` function, I have to use the `context`. But I can still call `write_all` using the `context`.

But `output` doesn't have access to `cr()`, which seems like is an output function to me.

Would it be simpler just to always use `context`, and remove the `output` capture?

Reply:
Sure thing.